runtime/lua/vim/fcnotify.lua

```diff
@@ -0,0 +1,224 @@
+-- Create a file watcher, each watcher is identified by the name of the file that it
+-- watches.
+
+local uv = vim.loop
+local Watcher = {}
+local w = {}
+local WatcherList = {}
+local commence = nil
+
+--- Callback for the check handle, checks if there are pending notifications
+--- for any watcher, and handles them as per the value of the `fcnotify`
+--- option.
+local function check_notifications()
+  for _, watcher in pairs(WatcherList) do
+    local option = vim.api.nvim_buf_get_option(watcher.bufnr, 'filechangenotify')
+    if watcher.pending_notifs and watcher.paused == false and option ~= 'off' then
+      if uv.fs_stat(watcher.ffname) ~= nil then
+        vim.api.nvim_command('checktime '..watcher.bufnr)
+        watcher.pending_notifs = false
+      else
+        error("ERR: File "..watcher.fname.." removed")
+      end
+    end
+  end
+end
+
+--- Checks if a buffer should have a watcher attached to it.
+--- Ignores all the buffers that aren't listed, or have a buf id
+--- less than 0.
+--- Ignores the [NO NAME] buffer.
+---
+--@param fname: The validity of this buffer will be checked.
+local function valid_buf(fname)
+  if fname == '' then
+    return false
+  end
+
+  local bufnr = vim.api.nvim_call_function('bufnr', {fname})
+
+  if bufnr < 0 then
+    return false
+  end
+
+  local buflisted = vim.api.nvim_buf_get_option(bufnr, 'buflisted')
+  local buftype = vim.api.nvim_buf_get_option(bufnr, 'buftype')
+
+  if not buflisted or buftype == 'nofile' or buftype == 'quickfix' then
+    return false
+  end
+  return true
+end
+
+local check_handle = uv.new_check()
+check_handle:start(vim.schedule_wrap(check_notifications))
+
+--- Creates and initializes a new watcher object with the given filename.
+---
+--@param fname: (required, string) The path that the watcher should watch.
+function Watcher:new(fname)
+  vim.validate{fname = {fname, 'string', false}}
+  -- get full path name for the file
+  local ffname = vim.api.nvim_call_function('fnamemodify', {fname, ':p'})
+  w = {bufnr = vim.api.nvim_call_function('bufnr', {fname}),
+       fname = fname, ffname = ffname, handle = nil,
+       paused = false, pending_notifs = false}
+  setmetatable(w, self)
+  self.__index = self
```

Comment:
You can set this just once abovve. I recommend moving definition of `Watcher` to line 55-ish. And then you can do `Watcher.__index = Watcher`

Reply:
Thanks! And done.